src/libraries/Common/src/Microsoft/Win32/SafeHandles/OverlappedValueTaskSource.Windows.cs

```diff
@@ -0,0 +1,216 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks.Sources;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>Reusable IValueTaskSource for Overlapped I/O.</summary>
+    internal abstract unsafe class OverlappedValueTaskSource : IValueTaskSource<int>, IValueTaskSource
+    {
+        private static readonly IOCompletionCallback s_ioCallback = IOCallback;
+
+        private readonly PreAllocatedOverlapped _preAllocatedOverlapped;
+        private readonly SafeHandle _fileHandle;
+        private readonly ThreadPoolBoundHandle _threadPoolBinding;
+        private readonly bool _canSeek;
+
+        internal MemoryHandle _memoryHandle;
+        protected int _bufferSize;
+        protected bool _isRead;
+        private ManualResetValueTaskSourceCore<int> _source; // mutable struct; do not make this readonly
+        private NativeOverlapped* _overlapped;
+        private CancellationTokenRegistration _cancellationRegistration;
+        /// <summary>
+        /// 0 when the operation hasn't been scheduled, non-zero when either the operation has completed,
+        /// in which case its value is a packed combination of the error code and number of bytes, or when
+        /// the read/write call has finished scheduling the async operation.
+        /// </summary>
+        internal ulong _result;
+
+        internal OverlappedValueTaskSource(SafeHandle fileHandle, ThreadPoolBoundHandle threadPoolBinding, bool canSeek)
+        {
+            _fileHandle = fileHandle;
+            _threadPoolBinding = threadPoolBinding;
+            _canSeek = canSeek;
+            _source.RunContinuationsAsynchronously = true;
+            _preAllocatedOverlapped = PreAllocatedOverlapped.UnsafeCreate(s_ioCallback, this, null);
```

Comment:
so far `SafeFileHandle` was using the `UnsafeCreate` (not touching the execution context):

https://github.com/dotnet/runtime/blob/5a62a7987525377e9aa870d866ef4cc073494c7b/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs#L66

while `PipeStream` used the "safe" ctor:

https://github.com/dotnet/runtime/blob/5a62a7987525377e9aa870d866ef4cc073494c7b/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs#L34

We should unify that and I am not sure which option to choose. If the safe ctor should be the default we need to add tests for this scenario (currently all are passing)

Reply:
Do we run any user code as part of s_ioCallback?  I believe the answer is "no", in which case it can be UnsafeCreate.